### ISSUE TITLE
We should not support SSLv2 in default SSL context

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3330]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#3330|
|Type|defect|
|Created|2008-07-03 13:06:50Z|
|Branch|https://github.com/twisted/twisted/tree/sslv2-3330|

SSLv2 has some security vulnerabilities, and SSLv23_METHOD, our current default method, is therefore vulnerable to degradation attacks in default configuration. At this point no one should be using SSLv2 anymore (SSL3 was released in 1996!). Firefox, for example, doesn't do SSLv2 anymore without asking user for lots of manual confirmation:

http://support.mozilla.com/eu/kb/Firefox+cannot+connect+securely+because+the+site+uses+an+older+insecure+version+of+the+SSL+protocol

My proposed solution:
Stick to SSLv23_METHOD, because it supports SSLv2, SSLv3 and TLSv1, but set SSL_OP_NO_SSLv2 so we don't actually use the insecure SSLv2. See:
http://www.openssl.org/docs/ssl/SSL_CTX_new.html

If people do want to support SSLv2 they can turn it on manually, but we should not have insecure defaults.

<details><summary>Searchable metadata</summary>

```
trac-id__3330 3330
type__defect defect
reporter__itamarst itamarst
priority__highest highest
milestone__ 
branch__branches_sslv2_3330 branches/sslv2-3330
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1215090410000000 1215090410000000
changetime__1230994151000000 1230994151000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

Raising priority cause it's a security issue which is the default in Twisted at the moment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [25761]) Branching to 'sslv2-3330'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

build results are at http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/sslv2-3330

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

It looks good, it would just be nice to add a few comments around the changes: explaining why we want to disable v2 for example. Also, explaining that SSLv23 actually includes TLSv1.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25767]) comment about SSLv23_METHOD and OP_NO_SSLv2

refs [#3330](https://github.com/twisted/twisted/issues/3330)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [25769]) Merge sslv2-3330

Author: exarkun
Reviewer: therve
Fixes: [#3330](https://github.com/twisted/twisted/issues/3330)

Drop SSLv2 support from `DefaultOpenSSLContextFactory`, leaving it supporting
only SSLv3 and TLSv1.  Add TLSv1 support to `DefaultContextFactory`, making it
support SSLv3 and TLSv1.

SSLv2 is insecure.  Allowing it at all makes a man in the middle attack possible.
SSLv3 has been around for long enough that it is no longer reasonable to every
need to interact with an SSLv2-only peer using one of the default context factories.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
